Btstrp from shub

GodloveD opened the pull request:
**Description of the Pull Request (PR):**

This PR allows users to bootstrap from shub using the following syntax in a def file:
```
Bootstrap: shub
From: author/image

%post
    echo "add more stuff here"
``` 
Just like the recent addition allowing users to bootstrap from an image on the local host, this addition will keep track of any previous definition files and store them with the metadata so that the users should be able to rebuild the container from scratch (assuming nothing important about the internet has changed 😜.  It will also warn the user if any env vars have changed during the bootstrap.   

**This fixes or addresses the following GitHub issues:**

- fixes Ref: #897 


**Checkoff for all PRs:**
✅  I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
✅   I have tested this PR locally with a `make test`
✅  This PR is NOT against the project's master branch
✅  I have added myself as a contributor to the [Author's file](https://github.com/singularityware/singularity/blob/master/AUTHORS.md)
✅  This PR is ready for review and/or merge


Attn: @singularityware-admin

gmkurtzer committed:
SquashFS/build support in progress

gmkurtzer committed:
Merge remote-tracking branch 'origin' into feature-squashbuild

gmkurtzer committed:
Merge branch 'development' into feature-squashbuild

gmkurtzer committed:
Allow a rebuild on an existing sandbox

gmkurtzer committed:
Conditional cookie on base OS build

gmkurtzer committed:
Merge branch 'development' into feature-squashbuild

gmkurtzer committed:
Be a bit smarter about the calling UID

gmkurtzer committed:
Fix variable names

gmkurtzer committed:
Removed slurm directory as it is causing `make maintainer-clean` errors

gmkurtzer committed:
Updates to `singularity build ...` and help

gmkurtzer committed:
Fix flawfinder error

GodloveD committed:
adding support for bootstrap from shub

GodloveD committed:
Adding shub bootstrap example

GodloveD committed:
added bootstrap from shub to documentation

GodloveD committed:
added test for bootstrap from shub

vsoch commented on libexec/bootstrap-scripts/main-deffile.sh:
```diff
@@ -50,7 +50,7 @@ if [ -n "${BOOTSTRAP:-}" -a -z "${SINGULARITY_BUILDNOBASE:-}" ]; then
 # take a snapshot of the environment for later comparison
-if [ ${BOOTSTRAP:-} = "localimage" ]; then
+if [ ${BOOTSTRAP:-} = "localimage" -o ${BOOTSTRAP:-} = "shub" ]; then
     SINGULARITY_STARTING_ENVIRONMENT=$(eval_abort env -i ${SINGULARITY_libexecdir}/singularity/helpers/record-env.sh ${SINGULARITY_ROOTFS})
```
ok cool, so this deals with environment / labels in the same way as a localimage?

GodloveD replied:
Yepper!  I think I might be able to add another `-o docker` and make it work in the same way for docker too.  But I haven't tried it yet.

gmkurtzer committed:
Support more build use cases

GodloveD committed:
first past read with some cleanup

GodloveD committed:
added functions for check filetypes

GodloveD committed:
added some tests for file types

gmkurtzer committed:
Lots of changes for the WIP of the image library cleanup

gmkurtzer committed:
Get rid of unneeded sources

gmkurtzer committed:
Minor cleanups to help outputs

gmkurtzer committed:
Updates to .gitignore

GodloveD committed:
still working on logic here

GodloveD committed:
Merge branch 'feature-squashbuild' into squash-from-localfile

GodloveD committed:
bugs be buggin

gmkurtzer committed:
Cleanups to loopdev bind code

gmkurtzer committed:
Cleanups to image

GodloveD committed:
added more checes to is_deffile

gmkurtzer committed:
Allow non-headerized image files to be understood

GodloveD committed:
removed $1 throughout

gmkurtzer committed:
Support for the image command class

gmkurtzer committed:
Updated tests for build

gmkurtzer committed:
Minor ordering change on image check (always check dirs first)

gmkurtzer committed:
Bind fix (loop device must be opened with right flags)

gmkurtzer committed:
image helper doesn't need to check for metadata folder

gmkurtzer committed:
Make sure we are mounting image as writable for import

gmkurtzer committed:
Updates to build

gmkurtzer committed:
Fix standard help output

gmkurtzer committed:
Fix wonky 'break;'

GodloveD committed:
starting on shub integration

GodloveD committed:
merging feature-squashbuild branch

gmkurtzer committed:
Minor usage fixes

gmkurtzer committed:
Minor modification to allow arbiturary commands for runscript

gmkurtzer committed:
Remove SINGULARITY image type mount requirement

gmkurtzer committed:
Minor fixes to environment files

gmkurtzer committed:
Lots of updates to tests for new image format

gmkurtzer committed:
Forgot to include squashfs-tools!

gmkurtzer committed:
Racecondition fix for directory based image

gmkurtzer committed:
Check for squashfs-tools

GodloveD committed:
Merge branch 'feature-squashbuild' into squash-from-localfile

gmkurtzer committed:
Minor fixup

gmkurtzer committed:
Fix minor typo of structure name

gmkurtzer committed:
Added debugging messages

gmkurtzer committed:
Fix bug in S_ISDIR return value

gmkurtzer committed:
Cleaning copyright entries for copy/pastes of header

gmkurtzer committed:
Pass any additional args to the helper

GodloveD committed:
Merge branch 'feature-squashbuild' into squash-from-localfile

GodloveD committed:
added support for build from image or from shub

GodloveD committed:
use full path

GodloveD committed:
bc is not a given (not installed on Ubuntu)

GodloveD committed:
reimplimenting bootstrap (for the time being)

GodloveD committed:
removed -F option from bootstrap (makes no sense now)

GodloveD committed:
adding some fixes that gmk suggested

gmkurtzer committed:
Removed old sources

gmkurtzer committed:
More robust image checking

GodloveD committed:
/bin/sh --> /bin/bash (at bequest of gmk)

GodloveD committed:
re-adding pointer to yanick

gmkurtzer committed:
Cleanups to mount

gmkurtzer committed:
Duplicate/malloc the return string

gmkurtzer committed:
Properly support writable images via init open_flags

gmkurtzer committed:
Support a persistent overlay image

gmkurtzer committed:
Updates for new singularity_image_init() options

gmkurtzer committed:
Initial support for persistent writable overlayFS image

GodloveD committed:
refactored to reduce redundant code

gmkurtzer committed:
Made the mount and final directories static per build

gmkurtzer committed:
Minor verbosity fixup

gmkurtzer committed:
Add global defines for CONTAINER_* paths

gmkurtzer committed:
Merge branch 'feature-squashbuild' into squash-from-localfile

cclerget committed:
Enforce pid namespace

cclerget committed:
Run cleanupd outside namespace to avoid /proc escape within PID namespace.
Open image before entering namespaces (see issues #714 #757)

cclerget committed:
Unlink trigger in case of failure (eg: set SINGULARITY_CLEANUPDIR=/tmp)

gmkurtzer committed:
Include a persistent overlay if one is requested in the mount

gmkurtzer committed:
Don't cap the overlay size limit if being called by root

gmkurtzer committed:
Properly define globals, and use those defines

GodloveD committed:
small bug

gmkurtzer committed:
Include appropriate option parsing for mount's overlay support

GodloveD committed:
properly emulating bootstrap now

GodloveD committed:
fixed bug in localimage driver and some others

cclerget committed:
Fix image_object segmentation fault when joining instance
and run required functions based on conditional statement

cclerget committed:
Fix sinit spawned 2 times and detach sinit from current session

GodloveD committed:
allow users to bootstrap to the same file over and over...

GodloveD committed:
renamed bootstrap for legacy behavior

GodloveD committed:
bugs found by Travis

GodloveD committed:
added docs and realized that I used an outdated version for bootstrap template

GodloveD committed:
adding docs and fixing bootstrap

gmkurtzer committed:
Merge branch 'feature-squashbuild-first' of https://github.com/cclerget/singularity into cclerget-feature-squashbuild-first

gmkurtzer committed:
Always init/open image, even when joining a running instance

gmkurtzer committed:
Merge branch 'cclerget-feature-squashbuild-first' into feature-squashbuild

gmkurtzer committed:
Supressing warning about no net namespace

gmkurtzer committed:
Clean/change arguments for start

gmkurtzer committed:
Fixed ordering of namespace separation

This fixes error with regards to net namespace not being able to open the FD during join

gmkurtzer committed:
Fixed error/warning messages about joining NS FD

gmkurtzer committed:
Conditionally call what functions are needed for join Vs. adhoc namespaces

GodloveD committed:
swapping cd for -C in tar commands

GodloveD committed:
fixing more things from review

gmkurtzer committed:
Merge branch 'squash-from-localfile' of https://github.com/GodloveD/singularity into GodloveD-squash-from-localfile

gmkurtzer committed:
Don't trust the user's SINGULARITY_bindir (or libexecdir)

gmkurtzer committed:
Merge branch 'GodloveD-squash-from-localfile' into feature-squashbuild

gmkurtzer committed:
Update instances to use the command class 'instance'

gmkurtzer committed:
Minor change to help

gmkurtzer committed:
Added an image command that returns the type of image it is

gmkurtzer committed:
Changed image binary name to image-type

gmkurtzer committed:
Test to make sure this is a writable Singularity image before modifying it

gmkurtzer committed:
Check to see if we have permission to modify image

gmkurtzer committed:
Updated is_image to use the new image-type backend binary

gmkurtzer committed:
Actually, we don't need any of it's output, just retval

GodloveD committed:
set label overwrite to True

vsoch commented on libexec/bootstrap-scripts/deffile-sections.sh:
```diff
@@ -167,7 +167,12 @@ if [ -z "${SINGULARITY_BUILDSECTION:-}" -o "${SINGULARITY_BUILDSECTION:-}" == "l
-                $ADD_LABEL --key "$KEY" --value "$VAL" --file "$SINGULARITY_ROOTFS/.singularity.d/labels.json"
+                if [ "${SINGULARITY_DEFFILE_BOOTSTRAP:-}" = "shub" -o "${SINGULARITY_DEFFILE_BOOTSTRAP:-}" = "localimage" ]; then
+                    $ADD_LABEL --key "$KEY" --value "$VAL" --file "$SINGULARITY_ROOTFS/.singularity.d/labels.json" -f "False"
+                else
```
Hey @GodloveD ! The force argument is a boolean, so you just need to add the flag (without any True/False), eg, here is the definition in python:
```
    parser.add_option('-f', dest="force",
                      help="force add (overwrite if exists)",
                      default=False, action='store_true')
```
and then it would just work like:

```
 $ADD_LABEL --key "$KEY" --value "$VAL" --file "$SINGULARITY_ROOTFS/.singularity.d/labels.json" -f
```

vsoch replied:
Did you try it with the "True" and does it error? Or just ignore it?

GodloveD replied:
Of course you are right.  duh.  Thanks.

vsoch replied:
haha no worries! There isn't clear documentation about this... it's all IN MY HEAD. :scream:

GodloveD committed:
derp

GodloveD committed:
adding a function to simplify build tests

GodloveD committed:
still adding tests and getting kinks worked out

GodloveD committed:
still working

GodloveD committed:
corrected some bugs with build from tar/tar.gz

GodloveD committed:
localimage can be a directory

GodloveD committed:
enforce runnning tests without privs

GodloveD committed:
rounding out tests a bit more

GodloveD committed:
quick commit so that i can change branches

GodloveD committed:
fixing a bug in Makefile.am with gmk

GodloveD committed:
fixed bug and removed test for feature yet to exist

gmkurtzer committed:
Update command groups to use spaces as well as dots

We should probably determine what we want to demonstrate in our
documentation. Both have pros and cons.

GodloveD committed:
classes to groups

GodloveD committed:
merging commands with spaces

GodloveD committed:
adding working args (except nobase which is now default behavior)

GodloveD committed:
adding check options into build

GodloveD committed:
cleaning up build help

GodloveD committed:
fixed command groups with spaces help docs

GodloveD committed:
trying to fix help tests

GodloveD committed:
fixing travis errors on help docs

GodloveD committed:
fixing travis tests

GodloveD committed:
fixing travis build errors

GodloveD committed:
fixing travis build errors

GodloveD committed:
Merge branch 'more-squashfs-tests' of https://github.com/GodloveD/singularity into more-squashfs-tests

GodloveD committed:
Merge branch 'feature-squashbuild' into more-squashfs-tests

yhcote committed:
Add prepheader program to add LAUNCH_STRING.

This adds the prepheader.c program and implements it in build.exec so
that each new image created now gets a launch string prepended to the
image.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Check for launch string in front of squash part.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
check for magic skipping header in simplier way

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Implement EXT3 image detection with or without launch string.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

yhcote committed:
Adding myself.

Signed-off-by: Yannick Cote <yhcote@gmail.com>

GodloveD committed:
added some instance tests

gmkurtzer committed:
Merge pull request #919 from yhcote/feature-squashbuild

Feature squashbuild

gmkurtzer committed:
Merge branch 'feature-squashbuild' into more-squashfs-tests

GodloveD committed:
pesky slash

GodloveD committed:
Merge branch 'more-squashfs-tests' of https://github.com/GodloveD/singularity into more-squashfs-tests

GodloveD committed:
pesky sudo

GodloveD committed:
adding wrapper to support legacy create function

GodloveD committed:
fixing license info

GodloveD committed:
adding test for non-existent build-spec

gmkurtzer committed:
Merge pull request #916 from GodloveD/squash-docs

Squash docs

gmkurtzer committed:
Merge branch 'more-squashfs-tests' of https://github.com/GodloveD/singularity into GodloveD-more-squashfs-tests

gmkurtzer committed:
Merge branch 'create-wrapper' of https://github.com/GodloveD/singularity into GodloveD-more-squashfs-tests

gmkurtzer committed:
Added src/prepheader to ignore

gmkurtzer committed:
Minor fixups to build.exec

gmkurtzer committed:
Slightly different method for calculating help

gmkurtzer committed:
Merge branch 'fix/the/thing' of https://github.com/vsoch/singularity into GodloveD-more-squashfs-tests

gmkurtzer committed:
Minor changes to action after merge

gmkurtzer committed:
Merge branch 'development' into GodloveD-more-squashfs-tests

GodloveD committed:
merging feature-squashbuild

GodloveD committed:
updated shub bootstrap for squashfs code

GodloveD committed:
adding test
